### PR TITLE
Allow kicking of dead players

### PR DESCRIPTION
**Pull request by tux314159**
Closes #20083.

This PR lets admins kick defeated players from the game. This is needed because some bad sports cause lag or reveal info about the game after they have lost, ruining the game for everyone else. ~~I also removed the no-kick-game-started thing because it doesn't seem to be very useful (there's no way for the admin to kick undefeated players anyway); this was needed as it was not letting the admin kick players even if they had lost.~~

EDIT: I updated it to allow kicking of defeated players too.

### CODE REVIEW

File: mods/cnc/chrome/ingame-infostats.yaml [hunk 1]
```diff
@@ -101,25 +102,35 @@ Container@SKIRMISH_STATS:
...
 							Height: 25
 							Shadow: True
 						Label@SCORE:
-							X: 360
+							X: 330
 							Width: 75
 							Height: 25
 							Align: Right
 							Shadow: True
+                        Button@KICK:
```
Comment (abcdefg30) [inline mods/cnc/chrome/ingame-infostats.yaml]: Spaces -> Tabs.

File: mods/common/chrome/ingame-infostats.yaml [hunk 2]
```diff
@@ -99,25 +99,35 @@ Container@SKIRMISH_STATS:
...
 							Height: 25
 							Shadow: True
 						Label@SCORE:
-							X: 360
+							X: 330
 							Width: 75
 							Height: 25
 							Align: Right
 							Shadow: True
+                        Button@KICK:
```
Comment (abcdefg30) [inline mods/common/chrome/ingame-infostats.yaml]: Same here.

File: mods/d2k/chrome/ingame-infostats.yaml [hunk 3]
```diff
@@ -80,7 +80,7 @@ Container@SKIRMISH_STATS:
 							Height: 20
 							Font: Bold
 							Align: Right
-				Container@PLAYER_TEMPLATE:
+                Container@PLAYER_TEMPLATE:
```
Comment (abcdefg30) [inline mods/d2k/chrome/ingame-infostats.yaml]: Here.

File: mods/d2k/chrome/ingame-infostats.yaml [hunk 4]
```diff
@@ -100,25 +100,35 @@ Container@SKIRMISH_STATS:
...
 							Height: 25
 							Shadow: True
 						Label@SCORE:
-							X: 360
+							X: 330
 							Width: 75
 							Height: 25
 							Align: Right
 							Shadow: True
+                        Button@KICK:
```
Comment (abcdefg30) [inline mods/d2k/chrome/ingame-infostats.yaml]: And here.

File: mods/cnc/chrome/ingame-infostats.yaml [hunk 5]
```diff
@@ -81,6 +81,7 @@ Container@SKIRMISH_STATS:
 							Height: 20
 							Font: Bold
 							Align: Right
+
```
Comment (abcdefg30) [inline mods/cnc/chrome/ingame-infostats.yaml]: Stray newline?

**Comment by tux314159** [comment]
(Closes #20083)

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 6]
```diff
@@ -766,11 +763,6 @@ static bool Kick(S server, Connection conn, Session.Client client, string s)
 				}
 
 				var kickClient = server.GetClient(kickConn);
-				if (server.State == ServerState.GameStarted && !kickClient.IsObserver)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:781]: The server should still verify that the client being kicked is either an observer or a defeated player.

Comment (tux314159) [reply OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:781]: This is already done in the chrome logic when deciding whether to show the kick button, is it still needed here?

Comment (abcdefg30) [reply OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:781]: Yes. Anyone can modify their client to send an order to kick players, so the server has to verify correctness.

Comment (tux314159) [reply OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:781]: Oh I see that makes sense, I'll add the checks then

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 7]
```diff
@@ -766,7 +766,13 @@ static bool Kick(S server, Connection conn, Session.Client client, string s)
...
+
+				var ws = server.worldPlayers
+					.Where(p => p != null && p.ClientIndex == kickClient.Index)
+					.Select(p => p.Outcome)
+					.Single();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: Are you sure this doesn't crash when trying to kick a spectator (which has no player associated I assume)?

File: OpenRA.Game/Server/Server.cs [hunk 8]
```diff
@@ -1065,9 +1062,9 @@ void InterpretServerOrder(Connection conn, Order o)
 							LobbyInfo.Slots = GameSave.Slots;
 
 							// Reassign clients to slots
-							//  - Bot ordering is preserved
-							//  - Humans are assigned on a first-come-first-serve basis
-							//  - Leftover humans become spectators
+							//	- Bot ordering is preserved
+							//	- Humans are assigned on a first-come-first-serve basis
+							//	- Leftover humans become spectators
```
Comment (abcdefg30) [inline OpenRA.Game/Server/Server.cs]: Should be reverted.

File: OpenRA.Game/Server/Server.cs [hunk 9]
```diff
@@ -75,7 +75,7 @@ public sealed class Server
 
 		ReplayRecorder recorder;
 		GameInformation gameInfo;
-		readonly List<GameInformation.Player> worldPlayers = new List<GameInformation.Player>();
+		public readonly List<GameInformation.Player> worldPlayers = new List<GameInformation.Player>();
```
Comment (abcdefg30) [inline OpenRA.Game/Server/Server.cs]:
```suggestion
		public readonly List<GameInformation.Player> WorldPlayers = new List<GameInformation.Player>();
```

**Review by abcdefg30 — COMMENTED** [review]
You need to add null checks for recorder in `OutOfSync` now too.

File: OpenRA.Game/Server/Server.cs (hunk 9 above)
Comment (tux314159) [reply OpenRA.Game/Server/Server.cs]: This is because of change of scope right

Comment (abcdefg30) [reply OpenRA.Game/Server/Server.cs]: Yes

File: OpenRA.Game/Server/Server.cs [hunk 10]
```diff
@@ -774,9 +774,11 @@ void OutOfSync(int frame)
 
 			// Make sure the written file is not valid
 			// TODO: storing a serverside replay on desync would be extremely useful
-			recorder.Metadata = null;
+			if (recorder != null) {
```
Comment (abcdefg30) [inline OpenRA.Game/Server/Server.cs]:
```suggestion
			if (recorder != null)
			{
```

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 11]
```diff
@@ -766,7 +766,15 @@ static bool Kick(S server, Connection conn, Session.Client client, string s)
...
 
 				var kickClient = server.GetClient(kickConn);
-				if (server.State == ServerState.GameStarted && !kickClient.IsObserver)
+
+				var ws = kickClient.IsObserver
+					? WinState.Lost
+					: server.WorldPlayers
+						.Where(p => p != null && p.ClientIndex == kickClient.Index)
+						.Select(p => p.Outcome)
+						.Single();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: We can combine this with the suggestions from Discord:
```suggestion
				var winState = server.WorldPlayers.SingleOrDefault(p => p?.ClientIndex == kickClient.Index)?.Outcome;
```
and then check if `winState` is `null` or not `WinState.Lost`.

**Review by PunkPun — COMMENTED** [review]
Crash
[crash.txt](https://github.com/OpenRA/OpenRA/files/9064675/crash.txt)

To reproduce it create 3 clients, 2 players and 1 spectator -> start game -> surrender as the non-admin player -> as admin kick the spectator -> crash on admin client, others are fine

**Comment by Punsho** [comment]
Consider aligning kick buttons
Screenshot 2022-07-07 at 17 19 05

**Comment by Punsho** [comment]
this pr will also collide with #20100 as we occupy the same UI space

**Comment by PunkPun** [comment]
Needs a rebase now that #20100 was merged

**Comment by tux314159** [comment]
What I need to do is: align the kick buttons right?

**Comment by PunkPun** [comment]
No, everything is aligned you should just rebase to upsteam/bleed

**Comment by tux314159** [comment]
Alright I've rebased on top of bleed

File: mods/common/chrome/ingame-infostats.yaml [hunk 12]
```diff
@@ -103,11 +103,11 @@ Container@SKIRMISH_STATS:
 							Template: ANONYMOUS_PLAYER_TOOLTIP
 						Label@NAME:
 							X: 29
-							Width: 191
+							Width: 150
```
Comment (PunkPun) [inline mods/common/chrome/ingame-infostats.yaml:106]: Revert every yaml change that isn't adding the kick button. #20100 already did all the changes for you

File: mods/common/chrome/ingame-infostats.yaml [hunk 13]
```diff
@@ -130,6 +130,16 @@ Container@SKIRMISH_STATS:
 							Checkmark: mute
 							Background: checkbox-toggle
 							TooltipContainer: TOOLTIP_CONTAINER
+						Button@KICK:
```
Comment (PunkPun) [inline mods/common/chrome/ingame-infostats.yaml:133]: Copy paste the `Button@KICK:` from line 175 to here. #20100 has changed how the button looks and works. 
Also do this in the other `ingame-infostats.yaml` files

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs [hunk 14]
```diff
@@ -162,6 +162,23 @@ public GameInfoStatsLogic(Widget widget, ModData modData, World world, OrderMana
 					muteCheckbox.IsVisible = () => !pp.IsBot && client.State != Session.ClientState.Disconnected && pp.ClientIndex != orderManager.LocalClient?.Index;
 					muteCheckbox.GetTooltipText = () => muteCheckbox.IsChecked() ? unmuteTooltip : muteTooltip;
 
+					var kickButton = item.Get<ButtonWidget>("KICK");
+					kickButton.IsVisible = () => Game.IsHost && client.Index != orderManager.LocalClient.Index && client.State != Session.ClientState.Disconnected && pp.WinState == WinState.Lost && !pp.IsBot;
+					kickButton.OnClick = () =>
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/GameInfoStatsLogic.cs:167]: Copy paste the `kickButton.OnClick` from line 215 to here. The way we define prompts has changed in #20142

**Comment by PunkPun** [comment]
Closing as superseded by #20521

**Pull request closed without merging**